### PR TITLE
Register plugin into registry

### PR DESCRIPTION
**Pull request by tagomoris**
To make straightforward dependencies from kinds of plugins to plugin.rb, and reduce duplicated implementation between plugin.rb and registry.rb.

### CODE REVIEW

File: lib/fluent/plugin.rb [hunk 1]
```diff
@@ -14,166 +14,133 @@
...
-        path = "fluent/plugin/filter_#{type}"
-      when 'buffer'
-        path = "fluent/plugin/buf_#{type}"
+    def self.new_impl(kind, registry, type)
+      # "'type' not found" is handled by registry
+      obj = registry.lookup(type)
+      case
+      when obj.is_a?(Class)
+        obj.new
+      when obj.respond_to?(:call)
```
Comment (repeatedly) [inline lib/fluent/plugin.rb]: The number of arity check is needed?

File: lib/fluent/registry.rb [hunk 2]
```diff
@@ -43,11 +46,18 @@ def lookup(type)
...
+      end
+      nil
+    end
+
     def search(type)
       path = "#{@search_prefix}#{type}"
 
       # prefer LOAD_PATH than gems
-      files = $LOAD_PATH.map { |lp|
+      files = ($LOAD_PATH + @paths).map { |lp|
```
Comment (repeatedly) [inline lib/fluent/registry.rb]: Use `(@paths + $LOAD_PATH)` instead.
Fluentd's plugin pathes should override default installed plugins.

File: lib/fluent/plugin.rb (hunk 1 above)
Comment (tagomoris) [inline lib/fluent/plugin.rb]: Correct. Better to have.

File: lib/fluent/registry.rb (hunk 2 above)
Comment (tagomoris) [inline lib/fluent/registry.rb]: Okay.

**Comment by tagomoris** [comment]
@repeatedly Can I merge after fixes of your points?

**Comment by repeatedly** [comment]
Yes.

@sonots Could you also check this?

File: lib/fluent/registry.rb (hunk 2 above)
Comment (tagomoris) [inline lib/fluent/registry.rb]: Ah, this code sorts found list and take last member of it to get latest version of files (it has another incorrectness, but it's out of scope for now).
https://github.com/fluent/fluentd/pull/838/files#diff-da556f72b0de83b0e659240bf9d879bbR66

So we should check files separately for `@paths` and `$LOAD_PATH`.

Comment (repeatedly) [inline lib/fluent/registry.rb]: okay

**Comment by sonots** [comment]
LGTM

**Comment by tagomoris** [comment]
OK. I'll merge this after green (Travis is already ok. Waiting appveyor).

**Comment by tagomoris** [comment]
An Appveyor task in a job is failing weird error and there's no way to rebuild it... I'll merge this :P

**Pull request merged**